File: app/src/main/java/com/yiann0s/retronebefits/MainActivity.java
package com.yiann0s.retronebefits;

import androidx.appcompat.app.AppCompatActivity;

import android.os.AsyncTask;
import android.os.Bundle;
import android.os.SystemClock;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.yiann0s.retronebefits.model.Car;
import com.yiann0s.retronebefits.utils.Service;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.ref.WeakReference;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;
import java.util.concurrent.ExecutionException;

import okhttp3.OkHttpClient;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    Long startTime, endTime;
    private final String BASE_URL = "https://navneet7k.github.io/";

    private final String TAG = "TEST.MainActivity";

    private TextView retrofitResultText, asyncTaskResultText;

    private ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.start_button).setOnClickListener(this);

        progressBar = findViewById(R.id.progress_bar);

        retrofitResultText = findViewById(R.id.retrofit_result);
        asyncTaskResultText = findViewById(R.id.asynctask_result);

    }

    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.start_button:
                retrofit(this);
                asyncTask();
                retrofitResultText.setVisibility(View.VISIBLE);
                asyncTaskResultText.setVisibility(View.VISIBLE);
                break;
            default:
                Toast.makeText(this,getString(R.string.error),Toast.LENGTH_SHORT).show();
        }
    }

    public void asyncTask(){
        MyAsyncTask task = new MyAsyncTask(this);

        String rr = "";
        try {
            rr = task.execute().get();
        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        try {
            JSONObject obj = new JSONObject(rr);
        } catch (JSONException e) {
            e.printStackTrace();
        }

    }


    private static class MyAsyncTask extends AsyncTask<String, Void, String> {

        //weak reference in the class scope
        private WeakReference<MainActivity> activityWeakReference;

        public MyAsyncTask(MainActivity activity) {
            activityWeakReference = new WeakReference<MainActivity>(activity);
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();

            //strong reference in method scope
            MainActivity activity = activityWeakReference.get();
            if ((activity == null) || (activity.isFinishing())) {
                return;
            }

            activity.startTime = SystemClock.elapsedRealtime();

            activity.progressBar.setVisibility(View.VISIBLE);
        }

        @Override
        protected String doInBackground(String... params) {
            //strong reference in method scope
            MainActivity activity = activityWeakReference.get();
            if ((activity == null) || (activity.isFinishing())) {
                return "";
            }

            HttpURLConnection urlConnection = null;
            String result = "";
            try {
                URL url = new URL("https://navneet7k.github.io/sample_array.json");
                urlConnection = (HttpURLConnection) url.openConnection();
                int code = urlConnection.getResponseCode();
                if (code == 200) {
                    InputStream in = new BufferedInputStream(urlConnection.getInputStream());
                    if (in != null) {
                        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(in));
                        String line = "";

                        activity.addDelay(4000);

                        while ((line = bufferedReader.readLine()) != null)
                            result += line;
                    }
                    in.close();
                }
                return result;
            } catch (MalformedURLException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                urlConnection.disconnect();
            }
            return result;
        }

        @Override
        protected void onPostExecute(String s) {
            super.onPostExecute(s);

            //strong reference in method scope
            MainActivity activity = activityWeakReference.get();
            if ((activity == null) || (activity.isFinishing())) {
                return;
            }

            activity.endTime = SystemClock.elapsedRealtime();
            activity.asyncTaskResultText.setText(activity.getResources().getString(R.string.time).replace("{x}",String.valueOf(activity.endTime-activity.startTime)).replace("{y}","AsyncTask"));
            activity.progressBar.setVisibility(View.INVISIBLE);
        }
    }

    private void addDelay(long delay){
        try {
            Thread.sleep(delay);
        } catch (InterruptedException e) {
            e.printStackTrace();
            Log.e(TAG, "addDelay: exception " + e.getMessage());
        }
    }

    public void retrofit(MainActivity activity){
        //weak reference in the class scope
        WeakReference<MainActivity> activityWeakReference = new WeakReference<MainActivity>(activity);

        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();


        Retrofit.Builder builder =
                new Retrofit.Builder()
                        .baseUrl(BASE_URL)
                        .addConverterFactory(
                                GsonConverterFactory.create()
                        );

        Retrofit retrofit =
                builder
                        .client(
                                httpClient.build()
                        )
                        .build();

        //is this good ???
        activityWeakReference.get().startTime= SystemClock.elapsedRealtime();
        Service service = retrofit.create(Service.class);
        Call<List<Car>> carList = service.getJson();
        carList.enqueue(new Callback<List<Car>>() {
            @Override
            public void onResponse(Call<List<Car>> call, Response<List<Car>> response) {
                //strong reference in method scope
                MainActivity activity = activityWeakReference.get();
                if ((activity == null) || (activity.isFinishing())) {
                    return;
                }

                if (response.isSuccessful() && response.body() != null ){

                    activity.endTime = SystemClock.elapsedRealtime();
                    activity.retrofitResultText.setText(activity.getResources().getString(R.string.time).replace("{x}",String.valueOf(activity.endTime-activity.startTime)).replace("{y}","Retrofit"));
                } else {
                    Log.d(TAG, "onResponse: reponse body isEmpty :" + (TextUtils.isEmpty(response.toString())));
                    Toast.makeText(activity,"Retrofiit:"+activity.getResources().getString(R.string.error),Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<Car>> call, Throwable t) {
                //strong reference in method scope
                MainActivity activity = activityWeakReference.get();
                if ((activity == null) || (activity.isFinishing())) {
                    return;
                }

                Toast.makeText(activity,"Exception: " + t.getMessage(),Toast.LENGTH_SHORT).show();
                if (t instanceof IOException) {
                    Log.d(TAG, "this is an actual network failure :( inform the user and possibly retry");
                } else {
                    Log.d(TAG, "conversion issue! big problems :(");
                }
            }
        });
    }

}

File: app/src/main/java/com/yiann0s/retronebefits/MainActivity.kt
package com.yiann0s.retronebefits

import android.os.AsyncTask
import android.os.Bundle
import android.os.SystemClock
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.yiann0s.retronebefits.model.Car
import com.yiann0s.retronebefits.utils.Service
import okhttp3.OkHttpClient
import org.json.JSONException
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.*
import java.lang.ref.WeakReference
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.concurrent.ExecutionException

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private var BASE_URL = "https://navneet7k.github.io/"
    private var TAG = "TEST.MainActivity"
    private var startTime : Long? = null
    private var endTime : Long? = null
    private var progressBar : ProgressBar? = null
    private var retrofitResultText : TextView? = null
    private var asyncTaskResultText : TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        (findViewById<Button>(R.id.start_button)).setOnClickListener(this)

        progressBar = findViewById(R.id.progress_bar)

        retrofitResultText = findViewById(R.id.retrofit_result)
        asyncTaskResultText = findViewById(R.id.asynctask_result)
    }

    override fun onClick(v: View?) {
        if (v?.id == R.id.start_button){
            retrofit(this)
            asyncTask()
            retrofitResultText?.visibility =View.VISIBLE
            asyncTaskResultText?.visibility = View.VISIBLE
        } else {
            Toast.makeText(this,getString(R.string.error),Toast.LENGTH_SHORT).show()
        }
    }

    fun asyncTask(){
        val task = MyAsyncTask(this)

        var rr = ""
        try {
            rr = task.execute().get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        try {
            JSONObject(rr)
        } catch (e : JSONException) {
            e.printStackTrace()
            Log.i(TAG, "asyncTask: exception " + e.message)
        }

    }

    companion object{
        class MyAsyncTask internal constructor( context: MainActivity? ) : AsyncTask<String, Void, String>() {

            private var response : String? = null
            //weak reference in the class scope
            private val activityWeakReference : WeakReference<MainActivity> = WeakReference(context)

            @Override
            override fun onPreExecute() {
                super.onPreExecute()

                //strong reference in method scope
                val context = activityWeakReference.get()
                if ((context == null) || (context.isFinishing)) {
                    return
                }

                context.startTime = SystemClock.elapsedRealtime()

                context.progressBar?.visibility = View.VISIBLE
            }

            override fun doInBackground(vararg p0 : String?):String {
                //strong reference in method scope
                val context = activityWeakReference?.get()
                if ((context == null) || (context.isFinishing)) {
                    return ""
                }

                var urlConnection : HttpURLConnection? = null

                try {
                    val url = URL("https://navneet7k.github.io/sample_array.json")
                    urlConnection = url.openConnection() as HttpURLConnection
                    val code: Int = urlConnection.responseCode
                    if (code == 200) {
                        val inputStreamReader = BufferedInputStream(urlConnection.inputStream)
                        val bufferedReader = BufferedReader(InputStreamReader(inputStreamReader))
                        var line: String?

                        if (App.HAS_DELAY){
                            context.addDelay(4000)
                        }

                        while (true){
                            line = bufferedReader.readLine()
                            if (line == null){
                                break
                            }
                            response += line
                        }
                        inputStreamReader.close()
                    }
                    return response!!
                } catch ( e : MalformedURLException) {
                    e.printStackTrace()
                } catch (e : IOException) {
                    e.printStackTrace()
                } finally {
                    urlConnection?.disconnect()
                }
                return response!!
            }

            override fun onPostExecute(s : String) {
                super.onPostExecute(s)

                //strong reference in method scope
                val context = activityWeakReference?.get()
                if ((context == null) || (context.isFinishing)) {
                    return
                }

                context.endTime = SystemClock.elapsedRealtime()
                val totalTimeLong = context.endTime!!- context.startTime!!
                val totalTime = "$totalTimeLong"
                context.asyncTaskResultText?.text = context.resources?.getString(R.string.time,  "AsyncTask", totalTime)
                context.progressBar?.visibility = View.INVISIBLE
            }
        }
    }

    fun addDelay(delay: Long){
        try {
            Thread.sleep(delay)
        } catch ( e : InterruptedException) {
            e.printStackTrace()
            Log.e(TAG, "addDelay: exception " + e.message)
        }
    }

    fun retrofit( context : MainActivity?){
        //weak reference in the class scope
        val activityWeakReference = WeakReference<MainActivity>(context)

        val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        //is this good ???
        activityWeakReference.get()?.startTime= SystemClock.elapsedRealtime()
        val service = retrofit.create(Service::class.java)
        val carList : Call<MutableList<Car>> = service.getJson()
        carList.enqueue( object : Callback<MutableList<Car>> {
            override fun onFailure(call: Call<MutableList<Car>>, t: Throwable) {
                val context: MainActivity? = activityWeakReference.get()
                if ((context == null) || (context.isFinishing)) {
                    return
                }

                Toast.makeText(context,"Exception: " + t.message,Toast.LENGTH_SHORT).show()
                if ( t is IOException) {
                    Log.d(TAG, "this is an actual network failure :( inform the user and possibly retry")
                } else {
                    Log.d(TAG, "conversion issue! big problems :(")
                }
            }

            override fun onResponse(
                    call: Call<MutableList<Car>>,
                    response: Response<MutableList<Car>>
            ) {
                val context: MainActivity? = activityWeakReference.get()
                if ((context == null) || (context.isFinishing)) {
                    return
                }

                if (response.isSuccessful && response.body() != null ){

                    context.endTime = SystemClock.elapsedRealtime()
                    val totalTimeLong = context.endTime!!- context.startTime!!
                    val totalTime = "$totalTimeLong"
                    context.retrofitResultText?.text = context.resources.getString(R.string.time, "Retrofit",totalTime)
                } else {
                    Log.d(TAG, "onResponse: reponse body isEmpty :" + (TextUtils.isEmpty(response.toString())))
                    Toast.makeText(context,"Retrofiit:"+context.resources.getString(R.string.error),Toast.LENGTH_SHORT).show()
                }
            }

        })
    }

}